### ISSUE TITLE
p2p authenticate peer on handshake

### ISSUE THREAD

**Issue by pmesnier** [EOSIO/eos#505]
One of the security enhancements identified as part of #291. Jonathan G. to support.

**Comment by jgiszczak** [comment]
I intend to use boost::asio's native support for OpenSSL to achieve this.  It will support peer authentication using any of a purely local certificate authority, a distributed private certificate authority, or a recognized global certificate authority.  This accommodates everything from testnets to loosely affiliated producers to contractually bound producers establishing a major chain.

Note that the example uses don't correspond to any particular scenario.  A testnet may be organized using certificates issued by a global CA and a major chain may be established with a private CA, at the discretion of the producers.

**Comment by jgiszczak** [comment]
Created cmakelists that generates a local certificate authority.

**Comment by bytemaster** [comment]
This is nice and all, but we need to authenticate peers against blockchain state which is the ultimate certificate authority. 

In particular, we will need to authenticate connections against producers.

I can see using OpenSSL to establish a secure encrypted connection, but authentication will have to be done by signing an ECDH shared secret using an EOS private key.

**Comment by jgiszczak** [comment]
Is a shared secret useful in some other context?  It seems an unnecessary vulnerability in this context.

If a single shared secret is compromised, an attacker could impersonate an arbitrary number of producer nodes in the network.  If instead a producer's companions are identified by their public keys, an attacker would have to compromise each producer's private key individually in order to impersonate them.

**Comment by jgiszczak** [comment]
Launcher now generates node keys and certificates.

**Comment by jgiszczak** [comment]
Revised how certificates were being generated.  Previous method was invalid.  Integration of certificate use is ongoing.

**Comment by jgiszczak** [comment]
Abandoning OpenSSL, much as it pains me.  It could be sufficiently automated that using an SSL-equipped connection would be relatively painless, but enabling that was invading more and more of the code.

Using producer keys at the application level instead.

**Comment by jgiszczak** [comment]
After a deep dive into the `fc` library, I have found the tools I need to do the job without reinventing any wheels.  I expect task completion within the next 24 hours.

**Comment by jgiszczak** [comment]
Somehow, some way, the key is unrecoverable.  Detailed debugging tomorrow.

**Comment by jgiszczak** [comment]
Pushed a patch which works, except on my system when one of the nodes has no producers, that node throws an exception trying to retrieve the genesis producers from the database.

**Comment by jgiszczak** [comment]
Blocked by #552.

**Comment by jgiszczak** [comment]
Pull request initiated.  All tests now pass.

**Comment by jgiszczak** [comment]
Updated launcher to generate the new config file options.  Now testing.

**Comment by jgiszczak** [comment]
Testing revealed some issues.  My favorite technique for sharing boost program options between plugins didn't work, so peer private keys had to get their own config file option.  Also reworked for ease of merging with ongoing sync work.  Testing now looking good.

**Comment by jgiszczak** [comment]
Pushed new pull request.

**Comment by jgiszczak** [comment]
Now supports authenticating producers, non-producing peers, and/or not authenticating. Authentication now occurs by public key rather than producer name. Protocol extended to eliminate replay attack: packet now contains a PGP-style message, hash, and hash signature, not just hash and hash signature. Only the peer configured to authenticate will check the handshake. If both sides are configured to authenticate, the connection is validated from both sides. ECDH shared secrets could not be used while supporting non-authenticated peers.

**Comment by jgiszczak** [comment]
Finished.
